src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs

```diff
@@ -34,9 +36,10 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
         private const int DefaultMaxRetries = 3;
         private int _maxRetries;
         private readonly HttpSource _httpSource;
-        private readonly ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>> _packageInfoCache =
-            new ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, AsyncLazy<HashSet<NuGetVersion>>> _packageVersionsCache =
+            new ConcurrentDictionary<string, AsyncLazy<HashSet<NuGetVersion>>>(StringComparer.OrdinalIgnoreCase);
         private readonly IReadOnlyList<Uri> _baseUris;
+        private string _chosenBaseUri;
```

Comment:
I don't like this, I would have liked to make it round-robin load balance across all the base URIs that the server provided in their service index.  But that would be a change in behavior which might risk breaking customers.  So, we re-use the base uri that the version list was obtained from, just like the old code.